modepy/modes.py

```diff
@@ -736,56 +785,270 @@ def symbolicize_function(f, dims, ref_coord_var_name="r"):
 # }}}
 
 
-# {{{ shape basis functions
+# {{{ basis interface
 
-# {{{ simplex
+class BasisNotOrthonormal(Exception):
+    pass
 
-@get_basis.register(Simplex)
-def _(shape: Simplex, order: int):
-    if shape.dims <= 3:
-        return simplex_onb(shape.dims, order)
-    else:
-        return simplex_monomial_basis(shape.dims, order)
 
+class Basis:
+    """
+    .. automethod:: orthonormality_weight(r)
+    .. autoattribute:: mode_ids
+    .. autoattribute:: functions
+    .. autoattribute:: gradients
+    """
 
-@get_grad_basis.register(Simplex)
-def _(shape: Simplex, order: int):
-    if shape.dims <= 3:
-        return grad_simplex_onb(shape.dims, order)
-    else:
-        return grad_simplex_monomial_basis(shape.dims, order)
+    def orthonormality_weight(self, rvec):
+        raise NotImplementedError
+
+    @property
+    def mode_ids(self):
+        """Return a tuple of of mode (basis function) identifiers, one for
+        each basis function. Mode identifiers should generally be viewed as opaque.
+        They are hashable. For some bases, they are tuples of length matching
+        the number of dimensions and indicating the order along each reference
+        axis.
+        """
+        raise NotImplementedError
+
+    @property
+    def functions(self):
+        """Return a tuple of (callable) basis functions of length matching
+        :attr:`mode_ids`.  Each function takes a vector of (r,s,t) reference
+        coordinates as input.
+        """
+        raise NotImplementedError
+
+    @property
+    def gradients(self):
+        """Return a tuple of basis functions of length matching :attr:`mode_ids`.
+        Each function takes a vector of (r,s,t) reference coordinates as input.
+        """
+        raise NotImplementedError
 
+# }}}
+
+
+# {{{ space-based basis retrieval
+
+@singledispatch
+def basis_for_space(space: FunctionSpace) -> Basis:
```

Comment:
Should this also take a shape as an argument? I'm thinking that something like inducer/modepy#15 needs a PN basis on the square.

Reply:
I think you're right. At first, I thought, "A P^N basis is a P^N basis", which is true, but its Vdm conditioning is going to depend on geometry.

:+1: for passing a shape here, too.

Reply:
Added in 97abd45

Reply:
Thanks, LGTM!